patches/api/0320-Add-Git-information-to-version-command-on-startup.patch

```diff
@@ -0,0 +1,154 @@
++
++    /**
++     * Get collection containing all META-INF/MANIFEST.MF streams
++     *
++     * @throws IOException if unable to get resources from classloader
++     */
++    public @NotNull Collection<@NotNull InputStream> getManifestStreams() throws IOException {
++        final Enumeration<URL> resources = getClass().getClassLoader().getResources("META-INF/MANIFEST.MF");
++        final Collection<InputStream> streams = new ArrayList<>(1);
++        while (resources.hasMoreElements()) {
++            streams.add(resources.nextElement().openStream());
++        }
++        return streams;
++    }
++
++    // Based on:
++    // https://github.com/jcabi/jcabi-manifests/blob/c4e1dd22bb6099769b8d279ebe3737e5b638b278/src/main/java/com/jcabi/manifests/Manifests.java#L209-L225
++    // https://github.com/jcabi/jcabi-manifests/blob/c4e1dd22bb6099769b8d279ebe3737e5b638b278/src/main/java/com/jcabi/manifests/Manifests.java#L381-L388
++
++    /**
++     * Get map containing entries from all manifests
++     *
++     * @throws IOException if unable to get manifest streams
++     */
++    public @NotNull Map<@NotNull String, @NotNull String> getManifestMap() throws IOException {
++        final HashMap<String, String> attribs = new HashMap<>();
++        for (final InputStream stream : getManifestStreams()) {
++            for (final Map.Entry<Object, Object> attr: new Manifest(stream).getMainAttributes().entrySet()) {
++                attribs.put(attr.getKey().toString(), attr.getValue().toString());
++            }
++        }
++        return attribs;
++    }
++}
+diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
+index e8414592b3afeb1e5db2b817b8fb7c13e073b9aa..6cc864af6cd5cdaa032986f4ebe7de7199646e62 100644
+--- a/src/main/java/org/bukkit/Bukkit.java
++++ b/src/main/java/org/bukkit/Bukkit.java
+@@ -50,6 +50,7 @@ import org.bukkit.util.CachedServerIcon;
+ import org.jetbrains.annotations.Contract;
+ import org.jetbrains.annotations.NotNull;
+ import org.jetbrains.annotations.Nullable;
++import io.papermc.paper.util.JarManifests; // Paper
+ 
+ /**
+  * Represents the Bukkit core, for version and Server singleton handling
+@@ -85,7 +86,22 @@ public final class Bukkit {
+         }
+ 
+         Bukkit.server = server;
+-        server.getLogger().info("This server is running " + getName() + " version " + getVersion() + " (Implementing API version " + getBukkitVersion() + ")");
++        // Paper start - add git information
++        server.getLogger().info(getVersionMessage());
++    }
++    /**
++      * Get message describing the version the server is running on.
++      */
++    public static String getVersionMessage() {
++        Map<String, String> attributes = JarManifests.MANIFEST_MAP;
++        @NotNull String gitBranch = attributes.get("Git-Branch");
++        @NotNull String gitCommit = attributes.get("Git-Commit");
```

Comment:
These are TECHNICALLY Nullable, but should always exist in paper jars. To be safe I did reverse comparisons below as to avoid potential NPE, but I'm not sure if we want to maybe set those to `unknown` or something, if they are null?